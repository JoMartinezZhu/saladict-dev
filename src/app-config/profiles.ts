import { DeepReadonly } from '@/typings/helpers'
import { genUniqueKey } from '@/_helpers/uniqueKey'
import { getALlDicts } from './dicts'

export type MtaAutoUnfold = '' | 'once' | 'always' | 'popup'

export type ProfileMutable = ReturnType<typeof _getDefaultProfile>
export type Profile = DeepReadonly<ProfileMutable>

export interface ProfileID {
  id: string
  name: string
}

export type ProfileIDList = Array<ProfileID>

export const getDefaultProfile: (id?: string) => Profile = _getDefaultProfile

export default getDefaultProfile

export function _getDefaultProfile (id?: string) {
  return {
    id: id || genUniqueKey(),

    /** auto unfold multiline textarea search box */
    mtaAutoUnfold: '' as MtaAutoUnfold,

    /** show waveform control panel */
    waveform: false,

    dicts: {
      /** default selected dictionaries */
      selected: [
        'bing',
        'cobuild',
        'cambridge',
        'youdao',
        'urban',
        'vocabulary',
        'google',
        'sogou',
        'zdic',
        'guoyu',
        'liangan',
        'googledict',
      ] as Array<keyof ReturnType<typeof getALlDicts>>,
      // settings of each dict will be auto-generated
      all: getALlDicts()
    },
  }
}

export function getDefaultProfileID (id?: string): ProfileID {
  return {
    id: id || genUniqueKey(),
    name: '%%_default_%%',
  }
}

export interface ProfileStorage {
  idItem: ProfileID
  profile: Profile
}

export function genProfilesStorage (): {
  profileIDList: ProfileIDList
  profiles: Profile[]
} {
  const defaultID = getDefaultProfileID()
  const defaultProfile = getDefaultProfile(defaultID.id)
  const sentenceStorage = sentence()
  const translationStorage = translation()
  const scholarStorage = scholar()

  return {
    profileIDList: [
      defaultID,
      sentenceStorage.idItem,
      translationStorage.idItem,
      scholarStorage.idItem,
    ],
    profiles: [
      defaultProfile,
      sentenceStorage.profile,
      translationStorage.profile,
      scholarStorage.profile,
    ]
  }
}

export function sentence (): ProfileStorage {
  const idItem = getDefaultProfileID()
  idItem.name = '%%_sentence_%%'

  const profile = getDefaultProfile(idItem.id) as ProfileMutable
  profile.dicts.selected = ['jukuu', 'bing', 'cnki', 'eudic', 'cobuild', 'cambridge', 'longman', 'macmillan']

  const allDict = profile.dicts.all
  allDict.bing.options.tense = false
  allDict.bing.options.phsym = false
  allDict.bing.options.cdef = false
  allDict.bing.options.related = false
  allDict.bing.options.sentence = 9999
  allDict.cnki.options.dict = false
  allDict.eudic.options.resultnum = 9999
  allDict.macmillan.options.related = false
  allDict.longman.options.wordfams = false
  allDict.longman.options.collocations = false
  allDict.longman.options.grammar = false
  allDict.longman.options.thesaurus = false
  allDict.longman.options.examples = true
  allDict.longman.options.bussinessFirst = false
  allDict.longman.options.related = false

  return { idItem, profile }
}

export function scholar (): ProfileStorage {
  const idItem = getDefaultProfileID()
  idItem.name = '%%_scholar_%%'

  const profile = getDefaultProfile(idItem.id) as ProfileMutable
  profile.dicts.selected = ['googledict', 'cambridge', 'cobuild', 'etymonline', 'cnki', 'macmillan', 'oald', 'websterlearner', 'google', 'sogou', 'zdic', 'guoyu', 'liangan']

  const allDict = profile.dicts.all
  allDict.macmillan.defaultUnfold = {
    english: false,
    chinese: false,
    japanese: false,
    korean: false,
    french: false,
    spanish: false,
    deutsch: false,
    others: false,
  }
  allDict.oald.defaultUnfold = {
    english: false,
    chinese: false,
    japanese: false,
    korean: false,
    french: false,
    spanish: false,
    deutsch: false,
    others: false,
  }
  allDict.websterlearner.defaultUnfold = {
    english: false,
    chinese: false,
    japanese: false,
    korean: false,
    french: false,
    spanish: false,
    deutsch: false,
    others: false,
  }
  allDict.google.selectionWC.min = 5
  allDict.sogou.selectionWC.min = 5

  return { idItem, profile }
}

export function translation (): ProfileStorage {
  const idItem = getDefaultProfileID()
  idItem.name = '%%_translation_%%'

  const profile = getDefaultProfile(idItem.id) as ProfileMutable
  profile.dicts.selected = ['google', 'tencent', 'sogou', 'baidu', 'caiyun', 'youdao', 'zdic', 'guoyu', 'liangan']
  profile.mtaAutoUnfold = 'always'

  return { idItem, profile }
}
