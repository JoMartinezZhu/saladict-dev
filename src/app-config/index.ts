import { DeepReadonly } from '@/typings/helpers'
import { getALlDicts } from './dicts'
import { getAllContextMenus } from './context-menus'
import { MtaAutoUnfold as _MtaAutoUnfold, _getDefaultProfile } from './profiles'
import { SupportedLangs } from '@/_helpers/lang-check'

export type LangCode = 'zh-CN' | 'zh-TW' | 'en'

const langUI = (browser.i18n.getUILanguage() || 'en')
const langCode: LangCode = /^zh-CN|zh-TW|en$/.test(langUI)
  ? langUI === 'zh-HK'
    ? 'zh-TW'
    : langUI as LangCode
  : 'en'

export type DictConfigsMutable = ReturnType<typeof getALlDicts>
export type DictConfigs = DeepReadonly<DictConfigsMutable>
export type DictID = keyof DictConfigsMutable
export type MtaAutoUnfold = _MtaAutoUnfold

export const enum TCDirection {
  center,
  top,
  right,
  bottom,
  left,
  topLeft,
  topRight,
  bottomLeft,
  bottomRight,
}

export type InstantSearchKey = 'direct' | 'ctrl' | 'alt' | 'shift'

/** '' means no preload */
export type PreloadSource = '' | 'clipboard' | 'selection'

export type AllDicts = ReturnType<typeof getALlDicts>

export type AppConfigMutable = ReturnType<typeof _getDefaultConfig>
export type AppConfig = DeepReadonly<AppConfigMutable>

export const getDefaultConfig: () => AppConfig = _getDefaultConfig
export default getDefaultConfig

function _getDefaultConfig () {
  return {
    version: 12,

    /** activate app, won't affect triple-ctrl setting */
    active: true,

    /** enable Google analytics */
    analytics: true,

    /** disable selection on type fields, like input and textarea */
    noTypeField: false,

    /** use animation for transition */
    animation: true,

    /** language code for locales */
    langCode,

    /** panel width */
    panelWidth: 450,

    /** panel max height in percentage, 0 < n < 100 */
    panelMaxHeightRatio: 80,

    /** custom panel css */
    panelCSS: '',

    /** panel font-size */
    fontSize: 13,

    /** sniff pdf request */
    pdfSniff: true,
    /** URLs, [regexp.source, match_pattern] */
    pdfWhitelist: [] as [string, string][],
    /** URLs, [regexp.source, match_pattern] */
    // tslint:disable-next-line: no-unnecessary-type-assertion
    pdfBlacklist: [
      ['^(http|https)://[^/]*?cnki\.net(/.*)?$', '*://*.cnki.net/*'],
    ] as [string, string][],

    /** track search history */
    searhHistory: false,
    /** incognito mode */
    searhHistoryInco: false,

    /** open word editor when adding a word to notebook */
    editOnFav: true,

    /** Show suggestions when typing on search box */
    searchSuggests: true,

    /** when and how to search text */
    mode: {
      /** show pop icon first */
      icon: true,
      /** how panel directly */
      direct: false,
      /** double click */
      double: false,
      /** holding a key */
      holding: {
        shift: false,
        ctrl: false,
        meta: false,
      },
      /** cursor instant capture */
      instant: {
        enable: false,
        key: 'alt' as InstantSearchKey,
        delay: 600,
      },
    },

    /** when and how to search text if the panel is pinned */
    pinMode: {
      /** direct: on mouseup */
      direct: true,
      /** double: double click */
      double: false,
      /** holding a key */
      holding: {
        shift: false,
        ctrl: false,
        meta: false,
      },
      /** cursor instant capture */
      instant: {
        enable: false,
        key: 'alt' as InstantSearchKey,
        delay: 600,
      },
    },

    /** when and how to search text inside dict panel */
    panelMode: {
      /** direct: on mouseup */
      direct: false,
      /** double: double click */
      double: false,
      /** holding a key */
      holding: {
        shift: false,
        ctrl: false,
        meta: false,
      },
      /** cursor instant capture */
      instant: {
        enable: false,
        key: 'alt' as InstantSearchKey,
        delay: 600,
      },
    },

    /** when this is a quick search standalone panel running */
    qsPanelMode: {
      /** direct: on mouseup */
      direct: false,
      /** double: double click */
      double: false,
      /** holding a key */
      holding: {
        shift: false,
        ctrl: true,
        meta: false,
      },
      /** cursor instant capture */
      instant: {
        enable: false,
        key: 'alt' as InstantSearchKey,
        delay: 600,
      },
    },

    /** hover instead of click */
    bowlHover: true,

    /** double click delay, in ms */
    doubleClickDelay: 450,

    /** show panel when triple press ctrl */
    tripleCtrl: true,

    /** preload source */
    tripleCtrlPreload: 'clipboard' as PreloadSource,

    /** auto search when triple hit ctrl */
    tripleCtrlAuto: false,

    /** where should the dict appears */
    tripleCtrlLocation: TCDirection.center,

    /** should panel be in a standalone window */
    tripleCtrlStandalone: true,

    /** standalone panel height */
    tripleCtrlHeight: 600,

    /** resize main widnow to leave space to standalone window */
    tripleCtrlSidebar: '' as '' | 'left' | 'right',

    /** should standalone panel response to page selection */
    tripleCtrlPageSel: true,

    /** browser action panel preload source */
    baPreload: 'clipboard' as PreloadSource,

    /** auto search when browser action panel shows */
    baAuto: false,

    /**
     * browser action behavior
     * 'popup_panel' - show dict panel
     * 'popup_fav' - add selection to notebook
     * 'popup_options' - opten options
     * others are same as context menus
     */
    baOpen: 'popup_panel',

    /** context tranlate engines */
    ctxTrans: {
      google: true,
      tencent: true,
      sogou: true,
      baidu: true,
    } as { [id in DictID]: boolean },

    /** start searching when source containing the languages */
    language: {
      chinese: true,
      english: true,
      japanese: true,
      korean: true,
      french: true,
      spanish: true,
      deutsch: true,
      others: false,
    } as SupportedLangs,

    /** auto pronunciation */
    autopron: {
      cn: {
        dict: '' as DictID | '',
        list: ['zdic', 'guoyu'] as DictID[]
      },
      en: {
        dict: '' as DictID | '',
        list: [
          'bing',
          'cambridge',
          'cobuild',
          'eudic',
          'longman',
          'macmillan',
          'oald',
          'urban',
          'websterlearner',
          'youdao',
        ] as DictID[],
        accent: 'uk' as 'us' | 'uk'
      },
      machine: {
        dict: '' as DictID | '',
        list: [
          'google',
          'sogou',
          'tencent',
          'baidu',
          'caiyun',
        ],
        // play translation or source
        src: 'trans' as 'trans' | 'searchText',
      }
    },

    /** URLs, [regexp.source, match_pattern] */
    whitelist: [] as [string, string][],
    /** URLs, [regexp.source, match_pattern] */
    // tslint:disable-next-line: no-unnecessary-type-assertion
    blacklist: [
      ['^https://stackedit\.io(/.*)?$', 'https://stackedit.io/*']
    ] as [string, string][],

    contextMenus: {
      selected: [
        'view_as_pdf',
        'google_translate',
        'google_search',
        'google_page_translate',
        'youdao_page_translate'
      ],
      all: getAllContextMenus()
    }
  }
}
