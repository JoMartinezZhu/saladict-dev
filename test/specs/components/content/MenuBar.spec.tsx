import React from 'react'
import { shallow } from 'enzyme'

import MenuBar, { MenuBarProps } from '@/content/components/MenuBar'
import noop from 'lodash/noop'
import { getDefaultConfig } from '@/app-config'
import { getDefaultProfile, getDefaultProfileID } from '@/app-config/profiles'

describe('Component/content/MenuBar', () => {
  it('should render correctly', () => {
    const config = {
      ...getDefaultConfig(),
      ...getDefaultProfile(),
    }
    const props: MenuBarProps = {
      t: x => x,
      searchSuggests: true,
      isTripleCtrl: true,
      isFav: false,
      isPinned: false,
      searchHistory: [],
      activeDicts: [config.id],
      activeConfigID: config.id,
      profiles: [getDefaultProfileID()],
      searchBox: {
        index: 0,
        text: '',
      },
      isShowMtaBox: false,

      searchText: noop,
      shareImg: noop,
      closePanel: noop,
      handleDragAreaTouchStart: noop,
      handleDragAreaMouseDown: noop,
      requestFavWord: noop,
      panelPinSwitch: noop,
      searchBoxUpdate: noop,
    }
    expect(shallow(<MenuBar {...props}/>)).toMatchSnapshot()
  })

  it('should render correctly with fav and pin', () => {
    const config = {
      ...getDefaultConfig(),
      ...getDefaultProfile(),
    }
    const props: MenuBarProps = {
      t: x => x,
      searchSuggests: true,
      isTripleCtrl: true,
      isFav: true,
      isPinned: true,
      searchHistory: [],
      activeDicts: [config.id],
      activeConfigID: config.id,
      profiles: [getDefaultProfileID()],
      searchBox: {
        index: 0,
        text: '',
      },
      isShowMtaBox: true,

      searchText: noop,
      shareImg: noop,
      closePanel: noop,
      handleDragAreaTouchStart: noop,
      handleDragAreaMouseDown: noop,
      requestFavWord: noop,
      panelPinSwitch: noop,
      searchBoxUpdate: noop,
    }
    expect(shallow(<MenuBar {...props}/>)).toMatchSnapshot()
  })
})
