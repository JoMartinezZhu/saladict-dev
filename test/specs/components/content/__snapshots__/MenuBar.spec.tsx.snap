// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/content/MenuBar should render correctly 1`] = `
<header
  className="panel-MenuBar"
>
  <button
    className="panel-MenuBar_Btn-dir"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchHistoryBack
      </title>
      <path
        d="M 7.191 15.999 L 21.643 1.548 C 21.998 1.192 21.998 0.622 21.643 0.267 C 21.288 -0.089 20.718 -0.089 20.362 0.267 L 5.267 15.362 C 4.911 15.718 4.911 16.288 5.267 16.643 L 20.362 31.732 C 20.537 31.906 20.771 32 20.999 32 C 21.227 32 21.462 31.913 21.636 31.732 C 21.992 31.377 21.992 30.807 21.636 30.451 L 7.191 15.999 Z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn-dir"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchHistoryNext
      </title>
      <path
        d="M 25.643 15.362 L 10.547 0.267 C 10.192 -0.089 9.622 -0.089 9.267 0.267 C 8.911 0.622 8.911 1.192 9.267 1.547 L 23.718 15.999 L 9.267 30.451 C 8.911 30.806 8.911 31.376 9.267 31.732 C 9.441 31.906 9.676 32 9.904 32 C 10.132 32 10.366 31.913 10.541 31.732 L 25.636 16.636 C 25.992 16.288 25.992 15.711 25.643 15.362 Z"
      />
    </svg>
  </button>
  <input
    className="panel-MenuBar_SearchBox"
    key="search-box"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value=""
  />
  <div>
    <CSSTransition
      classNames="panel-MenuBar_SuggestPanel"
      in={false}
      timeout={100}
      unmountOnExit={true}
    />
  </div>
  <button
    className="panel-MenuBar_Btn"
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 52.966 52.966"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchText
      </title>
      <path
        d="M51.704 51.273L36.844 35.82c3.79-3.8 6.14-9.04 6.14-14.82 0-11.58-9.42-21-21-21s-21 9.42-21 21 9.42 21 21 21c5.082 0 9.747-1.817 13.383-4.832l14.895 15.49c.196.206.458.308.72.308.25 0 .5-.093.694-.28.398-.382.41-1.015.028-1.413zM21.984 40c-10.478 0-19-8.523-19-19s8.522-19 19-19 19 8.523 19 19-8.525 19-19 19z"
      />
    </svg>
  </button>
  <div
    className="panel-MenuBar_DragArea"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="panel-MenuBar_SettingsWrapper"
  >
    <button
      className="panel-MenuBar_Btn"
      disabled={false}
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <svg
        className="panel-MenuBar_Icon"
        height="30"
        viewBox="0 0 612 612"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          tipOpenSettings
        </title>
        <path
          d="M0 97.92v24.48h612V97.92H0zm0 220.32h612v-24.48H0v24.48zm0 195.84h612V489.6H0v24.48z"
        />
      </svg>
    </button>
    <CSSTransition
      classNames="panel-MenuBar_ProfilePanel"
      in={false}
      timeout={100}
      unmountOnExit={true}
    />
  </div>
  <button
    className="panel-MenuBar_Btn"
    disabled={true}
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon-fav "
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipAddToNotebook
      </title>
      <path
        d="M 23.363 2 C 20.105 2 17.3 4.65 16.001 7.42 C 14.701 4.65 11.896 2 8.637 2 C 4.145 2 0.5 5.646 0.5 10.138 C 0.5 19.278 9.719 21.673 16.001 30.708 C 21.939 21.729 31.5 18.986 31.5 10.138 C 31.5 5.646 27.855 2 23.363 2 Z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon-history"
      height="30"
      viewBox="0 0 64 64"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipOpenHistory
      </title>
      <path
        d="M34.688 3.315c-15.887 0-28.812 12.924-28.81 28.73-.012.25-.157 4.434 1.034 8.94l-3.88-2.262c-.965-.56-2.193-.235-2.76.727-.557.96-.233 2.195.728 2.755l9.095 5.302c.02.01.038.013.056.022.1.05.2.09.31.12.07.02.14.05.21.07.09.02.176.02.265.03.06.003.124.022.186.022.036 0 .07-.01.105-.015.034 0 .063.007.097.004.05-.003.097-.024.146-.032.097-.017.19-.038.28-.068.08-.028.157-.06.23-.095.086-.04.165-.085.24-.137.074-.046.14-.096.206-.15.07-.06.135-.125.198-.195.06-.067.11-.135.16-.207.026-.04.062-.07.086-.11.017-.03.017-.067.032-.1.03-.053.07-.1.096-.16l3.62-8.96c.417-1.03-.08-2.205-1.112-2.622-1.033-.413-2.207.083-2.624 1.115l-1.86 4.6c-1.24-4.145-1.1-8.406-1.093-8.523C9.92 18.455 21.04 7.34 34.7 7.34c13.663 0 24.78 11.116 24.78 24.78S48.357 56.9 34.694 56.9c-1.114 0-2.016.902-2.016 2.015s.9 2.02 2.012 2.02c15.89 0 28.81-12.925 28.81-28.81 0-15.89-12.923-28.814-28.81-28.814z"
      />
      <path
        d="M33.916 36.002c.203.084.417.114.634.13.045.002.09.026.134.026.236 0 .465-.054.684-.134.06-.022.118-.054.177-.083.167-.08.32-.18.463-.3.03-.023.072-.033.103-.07L48.7 22.98c.788-.79.788-2.064 0-2.852-.787-.788-2.062-.788-2.85 0l-11.633 11.63-10.44-4.37c-1.032-.432-2.208.052-2.64 1.08-.43 1.027.056 2.208 1.08 2.638L33.907 36c.002 0 .006 0 .01.002z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 53.011 53.011"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipPinPanel
      </title>
      <path
        className="panel-MenuBar_Icon-pin "
        d="M52.963 21.297c-.068-.33-.297-.603-.61-.727-8.573-3.416-16.172-.665-18.36.288L19.113 8.2C19.634 3.632 17.17.508 17.06.372c-.18-.22-.442-.356-.725-.372-.282-.006-.56.09-.76.292L.32 15.546c-.202.2-.308.48-.29.765.015.285.152.55.375.727 2.775 2.202 6.35 2.167 7.726 2.055l12.722 14.953c-.868 2.23-3.52 10.27-.307 18.337.124.313.397.54.727.61.067.013.135.02.202.02.263 0 .518-.104.707-.293l14.57-14.57 13.57 13.57c.196.194.452.292.708.292s.512-.098.707-.293c.39-.392.39-1.024 0-1.415l-13.57-13.57 14.527-14.528c.237-.238.34-.58.27-.91zm-17.65 15.458L21.89 50.18c-2.437-8.005.993-15.827 1.03-15.91.158-.352.1-.764-.15-1.058L9.31 17.39c-.19-.225-.473-.352-.764-.352-.05 0-.103.004-.154.013-.036.007-3.173.473-5.794-.954l13.5-13.5c.604 1.156 1.39 3.26.964 5.848-.058.346.07.697.338.924l15.785 13.43c.31.262.748.31 1.105.128.077-.04 7.378-3.695 15.87-1.017L35.313 36.754z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 31.112 31.112"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipClosePanel
      </title>
      <path
        d="M31.112 1.414L29.698 0 15.556 14.142 1.414 0 0 1.414l14.142 14.142L0 29.698l1.414 1.414L15.556 16.97l14.142 14.142 1.414-1.414L16.97 15.556"
      />
    </svg>
  </button>
</header>
`;

exports[`Component/content/MenuBar should render correctly with fav and pin 1`] = `
<header
  className="panel-MenuBar"
>
  <button
    className="panel-MenuBar_Btn-dir"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchHistoryBack
      </title>
      <path
        d="M 7.191 15.999 L 21.643 1.548 C 21.998 1.192 21.998 0.622 21.643 0.267 C 21.288 -0.089 20.718 -0.089 20.362 0.267 L 5.267 15.362 C 4.911 15.718 4.911 16.288 5.267 16.643 L 20.362 31.732 C 20.537 31.906 20.771 32 20.999 32 C 21.227 32 21.462 31.913 21.636 31.732 C 21.992 31.377 21.992 30.807 21.636 30.451 L 7.191 15.999 Z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn-dir"
    disabled={true}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchHistoryNext
      </title>
      <path
        d="M 25.643 15.362 L 10.547 0.267 C 10.192 -0.089 9.622 -0.089 9.267 0.267 C 8.911 0.622 8.911 1.192 9.267 1.547 L 23.718 15.999 L 9.267 30.451 C 8.911 30.806 8.911 31.376 9.267 31.732 C 9.441 31.906 9.676 32 9.904 32 C 10.132 32 10.366 31.913 10.541 31.732 L 25.636 16.636 C 25.992 16.288 25.992 15.711 25.643 15.362 Z"
      />
    </svg>
  </button>
  <input
    className="panel-MenuBar_SearchBox"
    key="search-box"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    type="text"
    value=""
  />
  <div>
    <CSSTransition
      classNames="panel-MenuBar_SuggestPanel"
      in={false}
      timeout={100}
      unmountOnExit={true}
    />
  </div>
  <button
    className="panel-MenuBar_Btn"
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 52.966 52.966"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipSearchText
      </title>
      <path
        d="M51.704 51.273L36.844 35.82c3.79-3.8 6.14-9.04 6.14-14.82 0-11.58-9.42-21-21-21s-21 9.42-21 21 9.42 21 21 21c5.082 0 9.747-1.817 13.383-4.832l14.895 15.49c.196.206.458.308.72.308.25 0 .5-.093.694-.28.398-.382.41-1.015.028-1.413zM21.984 40c-10.478 0-19-8.523-19-19s8.522-19 19-19 19 8.523 19 19-8.525 19-19 19z"
      />
    </svg>
  </button>
  <div
    className="panel-MenuBar_DragArea"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="panel-MenuBar_SettingsWrapper"
  >
    <button
      className="panel-MenuBar_Btn"
      disabled={false}
      onClick={[Function]}
      onKeyUp={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <svg
        className="panel-MenuBar_Icon"
        height="30"
        viewBox="0 0 612 612"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          tipOpenSettings
        </title>
        <path
          d="M0 97.92v24.48h612V97.92H0zm0 220.32h612v-24.48H0v24.48zm0 195.84h612V489.6H0v24.48z"
        />
      </svg>
    </button>
    <CSSTransition
      classNames="panel-MenuBar_ProfilePanel"
      in={false}
      timeout={100}
      unmountOnExit={true}
    />
  </div>
  <button
    className="panel-MenuBar_Btn"
    disabled={true}
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon-fav isActive"
      height="30"
      viewBox="0 0 32 32"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipAddToNotebook
      </title>
      <path
        d="M 23.363 2 C 20.105 2 17.3 4.65 16.001 7.42 C 14.701 4.65 11.896 2 8.637 2 C 4.145 2 0.5 5.646 0.5 10.138 C 0.5 19.278 9.719 21.673 16.001 30.708 C 21.939 21.729 31.5 18.986 31.5 10.138 C 31.5 5.646 27.855 2 23.363 2 Z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon-history"
      height="30"
      viewBox="0 0 64 64"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipOpenHistory
      </title>
      <path
        d="M34.688 3.315c-15.887 0-28.812 12.924-28.81 28.73-.012.25-.157 4.434 1.034 8.94l-3.88-2.262c-.965-.56-2.193-.235-2.76.727-.557.96-.233 2.195.728 2.755l9.095 5.302c.02.01.038.013.056.022.1.05.2.09.31.12.07.02.14.05.21.07.09.02.176.02.265.03.06.003.124.022.186.022.036 0 .07-.01.105-.015.034 0 .063.007.097.004.05-.003.097-.024.146-.032.097-.017.19-.038.28-.068.08-.028.157-.06.23-.095.086-.04.165-.085.24-.137.074-.046.14-.096.206-.15.07-.06.135-.125.198-.195.06-.067.11-.135.16-.207.026-.04.062-.07.086-.11.017-.03.017-.067.032-.1.03-.053.07-.1.096-.16l3.62-8.96c.417-1.03-.08-2.205-1.112-2.622-1.033-.413-2.207.083-2.624 1.115l-1.86 4.6c-1.24-4.145-1.1-8.406-1.093-8.523C9.92 18.455 21.04 7.34 34.7 7.34c13.663 0 24.78 11.116 24.78 24.78S48.357 56.9 34.694 56.9c-1.114 0-2.016.902-2.016 2.015s.9 2.02 2.012 2.02c15.89 0 28.81-12.925 28.81-28.81 0-15.89-12.923-28.814-28.81-28.814z"
      />
      <path
        d="M33.916 36.002c.203.084.417.114.634.13.045.002.09.026.134.026.236 0 .465-.054.684-.134.06-.022.118-.054.177-.083.167-.08.32-.18.463-.3.03-.023.072-.033.103-.07L48.7 22.98c.788-.79.788-2.064 0-2.852-.787-.788-2.062-.788-2.85 0l-11.633 11.63-10.44-4.37c-1.032-.432-2.208.052-2.64 1.08-.43 1.027.056 2.208 1.08 2.638L33.907 36c.002 0 .006 0 .01.002z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 53.011 53.011"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipPinPanel
      </title>
      <path
        className="panel-MenuBar_Icon-pin isActive"
        d="M52.963 21.297c-.068-.33-.297-.603-.61-.727-8.573-3.416-16.172-.665-18.36.288L19.113 8.2C19.634 3.632 17.17.508 17.06.372c-.18-.22-.442-.356-.725-.372-.282-.006-.56.09-.76.292L.32 15.546c-.202.2-.308.48-.29.765.015.285.152.55.375.727 2.775 2.202 6.35 2.167 7.726 2.055l12.722 14.953c-.868 2.23-3.52 10.27-.307 18.337.124.313.397.54.727.61.067.013.135.02.202.02.263 0 .518-.104.707-.293l14.57-14.57 13.57 13.57c.196.194.452.292.708.292s.512-.098.707-.293c.39-.392.39-1.024 0-1.415l-13.57-13.57 14.527-14.528c.237-.238.34-.58.27-.91zm-17.65 15.458L21.89 50.18c-2.437-8.005.993-15.827 1.03-15.91.158-.352.1-.764-.15-1.058L9.31 17.39c-.19-.225-.473-.352-.764-.352-.05 0-.103.004-.154.013-.036.007-3.173.473-5.794-.954l13.5-13.5c.604 1.156 1.39 3.26.964 5.848-.058.346.07.697.338.924l15.785 13.43c.31.262.748.31 1.105.128.077-.04 7.378-3.695 15.87-1.017L35.313 36.754z"
      />
    </svg>
  </button>
  <button
    className="panel-MenuBar_Btn"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      className="panel-MenuBar_Icon"
      height="30"
      viewBox="0 0 31.112 31.112"
      width="30"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        tipClosePanel
      </title>
      <path
        d="M31.112 1.414L29.698 0 15.556 14.142 1.414 0 0 1.414l14.142 14.142L0 29.698l1.414 1.414L15.556 16.97l14.142 14.142 1.414-1.414L16.97 15.556"
      />
    </svg>
  </button>
</header>
`;
